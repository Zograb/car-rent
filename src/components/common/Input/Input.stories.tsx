import type { Meta, StoryObj } from '@storybook/react';
import { CiSearch } from 'react-icons/ci';

import Input from './Input';

const meta: Meta<typeof Input> = {
  title: 'Input',
  component: Input,
  tags: ['autodocs'],
};

const icon = <CiSearch />;

type Story = StoryObj<typeof Input>;

export default meta;

export const Basic: Story = {};

export const WithIcon: Story = {
  args: {
    icon: <CiSearch />
  },
};

export const WithIconOnLeft: Story = {
  args: {
    icon,
    iconPosition: 'left',
  },
};

export const FullWidth: Story = {
  args: {
    icon,
    iconPosition: 'left',
    fullWidth: true,
  },
};
