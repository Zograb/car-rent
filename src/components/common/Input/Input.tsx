import type { ComponentPropsWithoutRef, FC, ReactNode } from 'react';
import type { CSSProp } from 'styled-components';
import tw from 'twin.macro';

export interface InputProps extends ComponentPropsWithoutRef<'input'> {
  // icon component
  icon?: ReactNode;

  // determines icon position on input
  iconPosition?: 'left' | 'right';

  // the input will occupy the entire width if true
  fullWidth?: boolean;

  // styles for input container
  containerStyles?: CSSProp;
}

const Input: FC<InputProps> = ({
  icon = null,
  iconPosition = 'right',
  fullWidth = false,
  containerStyles = tw``,
  ...otherProps
}) => (
  <div
    css={[
      tw`relative inline-block`,
      !fullWidth && tw`w-80`,
      fullWidth && tw`w-full`,
      containerStyles,
    ]}
    data-testid="input-container"
  >
    {iconPosition === 'left' && (
      <div
        tw="absolute top-1/2 left-4 -translate-y-1/2 text-3xl text-gray2"
        data-testid="icon-left"
      >
        {icon}
      </div>
    )}
    <input
      placeholder="Search here..."
      css={[
        tw`ease-linear duration-200 text-gray3 rounded-xl h-12 w-full outline-none`,
        tw`shadow-input focus:shadow-input-focused`,
        iconPosition === 'right' && tw`pl-8 pr-16`,
        iconPosition === 'left' && tw`pl-14 pr-8`,
      ]}
      {...otherProps}
    />
    {iconPosition === 'right' && (
      <div
        tw="absolute top-1/2 right-7 -translate-y-1/2 text-3xl text-gray2"
        data-testid="icon-right"
      >
        {icon}
      </div>
    )}
  </div>
);

export default Input;
