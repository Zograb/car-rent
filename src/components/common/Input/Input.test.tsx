import { render, screen } from '@testing-library/react';
import { BiSearch } from '@react-icons/all-files/bi/BiSearch';
import tw from 'twin.macro';

import { Input } from './Input';

describe('Input', () => {
  it('should match snapshot', () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });

  it('should render input', () => {
    render(<Input />);
    const input = screen.getByRole('textbox');
    expect(input).toBeInTheDocument();
  });

  it('should render input with icon', () => {
    render(<Input icon={<BiSearch />} />);
    const input = screen.getByRole('textbox');
    const inputIcon = screen.getByTestId('icon-right');
    expect(input).toBeInTheDocument();
    expect(inputIcon).toBeInTheDocument();
  });

  it('should render input with icon on left', () => {
    render(<Input icon={<BiSearch />} iconPosition="left" />);
    const input = screen.getByRole('textbox');
    const inputIcon = screen.getByTestId('icon-left');
    expect(input).toBeInTheDocument();
    expect(inputIcon).toBeInTheDocument();
  });

  it('should render full width input', () => {
    render(<Input fullWidth />);
    const inputContainer = screen.getByTestId('input-container');
    expect(inputContainer).toBeInTheDocument();
    expect(inputContainer).toHaveStyleRule(
      'width',
      tw`w-full`.width as jest.Value,
    );
  });
});

