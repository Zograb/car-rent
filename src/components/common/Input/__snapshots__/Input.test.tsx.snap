// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  width: 20rem;
}

.c1 {
  height: 3rem;
  width: 100%;
  border-radius: 0.75rem;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
  outline: 2px solid transparent;
  outline-offset: 2px;
  transition-duration: 200ms;
  transition-timing-function: linear;
  --tw-shadow: 0 2.7px 27px rgba(69, 69, 80, 0.1);
  --tw-shadow-colored: 0 2.7px 27px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
  padding-left: 2rem;
  padding-right: 4rem;
}

.c1:focus {
  --tw-shadow: 0 3.5px 30px rgba(69, 69, 80, 0.2);
  --tw-shadow-colored: 0 3.5px 30px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.c2 {
  position: absolute;
  right: 1.75rem;
  top: 50%;
  --tw-translate-y: -50%;
  transform: translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  font-size: 1.875rem;
  line-height: 2.25rem;
  --tw-text-opacity: 1;
  color: rgb(82 82 86 / var(--tw-text-opacity));
}

<div>
  <div
    class="c0"
    data-testid="input-container"
  >
    <input
      class="c1"
      placeholder="Search here..."
    />
    <div
      class="c2"
      data-testid="icon-right"
    />
  </div>
</div>
`;
